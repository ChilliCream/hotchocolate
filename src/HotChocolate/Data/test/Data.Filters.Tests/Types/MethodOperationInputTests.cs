using System.Linq;
using Snapshooter.Xunit;
using Xunit;

namespace HotChocolate.Data.Filters
{
    public class MethodOperationInputTests
    {
        [Fact]
        public void Create_Explicit_Operation()
        {
            // arrange
            var convention = new FilterConvention(x => x.UseMock());

            // act
            ISchema schema = SchemaBuilder.New()
                .AddQueryType<QueryExplicit>()
                .AddConvention<IFilterConvention>(convention)
                .UseFiltering()
                .Create();

            // assert
            schema.ToString().MatchSnapshot();
        }

        [Fact]
        public void Create_Implicit_Operation()
        {
            // arrange
            var convention = new FilterConvention(x => x.UseMock());

            // act
            ISchema schema = SchemaBuilder.New()
                .AddQueryType<Query>()
                .AddConvention<IFilterConvention>(convention)
                .UseFiltering()
                .Create();

            // assert
            schema.ToString().MatchSnapshot();
        }

        public class QueryExplicit
        {
            [UseFiltering(Type = typeof(FooOperationType))]
            public IQueryable<Foo> Foos() => new Foo[0].AsQueryable();
        }

        public class Query
        {
            [UseFiltering]
            public IQueryable<Foo> Foos() => new Foo[0].AsQueryable();
        }

        public class FooOperationType : FilterInputType<Foo>
        {
            protected override void Configure(IFilterInputTypeDescriptor<Foo> descriptor)
            {
                descriptor.Name("Test");
                descriptor.Operation(x => x.SimpleMethod()).Name("TestSimpleMethod");
                descriptor.Operation(x => x.ComplexMethod()).Name("TestComplexMethod");
            }
        }

        public class Foo
        {
            public bool SimpleMethod() => true;

            public Bar ComplexMethod() => new Bar();
        }

        public class Bar
        {
            public string StringOperation { get; set; } = "";
        }
    }
}