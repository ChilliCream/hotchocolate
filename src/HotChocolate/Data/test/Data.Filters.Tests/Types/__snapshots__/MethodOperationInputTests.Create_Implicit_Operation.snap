schema {
  query: Query
}

type Bar {
  stringOperation: String!
}

type Foo {
  complexMethod: Bar!
  simpleMethod: Boolean!
}

type Query {
  foos(where: FooFilter): [Foo!]!
}

input FooFilter {
  and: [FooFilter!]
  or: [FooFilter!]
}

"The `Boolean` scalar type represents `true` or `false`."
scalar Boolean

"The `String` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text."
scalar String
